src/Consolonia.Core/Drawing/DrawingContextImpl.Bitmap.cs

```diff
@@ -0,0 +1,395 @@
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Avalonia;
+using Avalonia.Media;
+using Avalonia.Media.Imaging;
+using Avalonia.Platform;
+using Consolonia.Core.Drawing.PixelBufferImplementation;
+using Consolonia.Core.Dummy;
+
+namespace Consolonia.Core.Drawing
+{
+    [StructLayout(LayoutKind.Sequential, Pack = 1)]
+    internal struct BgraColor
+    {
+        public byte B;
+        public byte G;
+        public byte R;
+        public byte A;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public BgraColor(byte b, byte g, byte r, byte a)
+        {
+            B = b;
+            G = g;
+            R = r;
+            A = a;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public readonly Color ToColor()
+        {
+            return Color.FromArgb(A, R, G, B);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static BgraColor FromColor(Color color)
+        {
+            return new BgraColor(color.B, color.G, color.R, color.A);
+        }
+
+        public static readonly BgraColor Transparent = new(0, 0, 0, 0);
+    }
+
+    internal partial class DrawingContextImpl
+    {
+        public void DrawBitmap(IBitmapImpl source, double opacity, Rect sourceRect, Rect destRect)
+        {
+            if (source is DummyBitmap)
+                return;
+
+            // resize bitmap to destination rect size
+            var targetRect = new Rect(Transform.Transform(new Point(destRect.TopLeft.X, destRect.TopLeft.Y)),
+                    Transform.Transform(new Point(destRect.BottomRight.X, destRect.BottomRight.Y)))
+                .ToPixelRect();
+
+            // Get the platform render interface to resize the bitmap
+            var renderInterface = AvaloniaLocator.Current.GetRequiredService<IPlatformRenderInterface>();
+
+            // Resize source to be target rect * 2 so we can map to quad pixels
+            var targetSize = new PixelSize(targetRect.Width * 2, targetRect.Height * 2);
+            using IBitmapImpl resizedBitmap =
+                renderInterface.ResizeBitmap(source, targetSize, BitmapInterpolationMode.MediumQuality);
+
+            // Get pixel data from the resized bitmap
+            var readableBitmap = (IReadableBitmapImpl)resizedBitmap;
+
+            using ILockedFramebuffer frameBuffer = readableBitmap.Lock();
+
+            int stride = frameBuffer.RowBytes;
+            int bytesPerPixel = frameBuffer.Format.BitsPerPixel / 8;
+            byte[] pixelBytes = new byte[stride * frameBuffer.Size.Height];
+            Marshal.Copy(frameBuffer.Address, pixelBytes, 0, pixelBytes.Length);
+            Span<BgraColor> pixels = MemoryMarshal.Cast<byte, BgraColor>(pixelBytes);
+
+            // this is reused by each pixel as we process the bitmap
+            Span<BgraColor> quadPixelColors = stackalloc BgraColor[4];
```

Comment:
Why do we need this pixels? It looks like a sub-array. Can't we create a span directly with a shift from original array?